pkg/storage/ingest/partition_offset_reader.go

```diff
@@ -138,10 +161,41 @@ func (p *partitionOffsetReader) FetchLastProducedOffset(ctx context.Context) (_
 		}
 	}()
 
+	offset, err := p.fetchPartitionOffset(ctx, kafkaEndOffset)
+	if err != nil {
+		return 0, err
+	}
+
+	// The offset we get is the offset at which the next message will be written, so to get the last produced offset
+	// we have to subtract 1. See DESIGN.md for more details.
+	return offset - 1, nil
+}
+
+// FetchPartitionStartOffset fetches and returns the start offset for a partition. This function returns 0 if no record has
+// been ever produced in the partition. This function issues a single request, but the Kafka client used under the
+// hood may retry a failed request until the retry timeout is hit.
+func (p *partitionOffsetReader) FetchPartitionStartOffset(ctx context.Context) (_ int64, returnErr error) {
+	startTime := time.Now()
+
+	p.partitionStartOffsetRequestsTotal.Inc()
+	defer func() {
+		// We track the latency also in case of error, so that if the request times out it gets
+		// pretty clear looking at latency too.
+		p.partitionStartOffsetLatency.Observe(time.Since(startTime).Seconds())
+
+		if returnErr != nil {
+			p.partitionStartOffsetFailuresTotal.Inc()
+		}
+	}()
+
+	return p.fetchPartitionOffset(ctx, kafkaStartOffset)
+}
+
+func (p *partitionOffsetReader) fetchPartitionOffset(ctx context.Context, position int64) (int64, error) {
```

Comment:
nit: `offset` is the kafka terminology

Reply:
The `position` value is specified in `Timestamp` field of `ListOffsetsRequestTopicPartition`. What we specify here is specifically the "start partition" or "end partition", reason why I've called it `position` instead of `offset`. Renaming it to `offset` or `timestamp` I think is more misleading than `position`. Thoughts?

Reply:
Sounds good. Thanks for putting the time to write it up